File: Aman_Bhaiya_DSA_List/Arrays/LQ36.java
package Aman_Bhaiya_DSA_List.Arrays;

public class LQ36 {
    // print all possible combinations of r elements in a n-size array

    public static void main(String[] args) {
        int[] arr = { 1, 2, 3, 4, 5 };
        int r = 3;
        int n = arr.length;
        printCombination(arr, n, r);
    }

    public static void printCombination(int[] arr, int n, int r) {
        int[] data = new int[r];
        combinationUtil(arr, data, 0, n - 1, 0, r);
    }

    public static void combinationUtil(int[] arr, int[] data, int start, int end, int index, int r) {
        if (index == r) {
            for (int j = 0; j < r; j++) {
                System.out.print(data[j] + " ");
            }
            System.out.println();
            return;
        }

        for (int i = start; i <= end && end - i + 1 >= r - index; i++) {
            data[index] = arr[i];
            combinationUtil(arr, data, i + 1, end, index + 1, r);
        }
    }
}

File: Aman_Bhaiya_DSA_List/Arrays/LQ36.py
# print all possible combinations of r elements in a n-size array

def combinationUtil(arr, n, r, index, data, i):
    if index == r:
        for j in range(r):
            print(data[j], end = " ")
        print()
        return

    if i >= n:
        return

    # current is included, put next at next location in data
    data[index] = arr[i]

    # current is included, put next at next location
    combinationUtil(arr, n, r, index + 1, data, i + 1)

    # current is excluded, replace it with next (Note that i+1 is passed, but index is not changed)
    combinationUtil(arr, n, r, index, data, i + 1)

def printCombination(arr, n, r):
    data = [0] * r
    combinationUtil(arr, n, r, 0, data, 0)

arr = [1, 2, 3, 4, 5]
r = 3
n = len(arr)
printCombination(arr, n, r)